### PR TITLE
[Python2] Handle urlparse'd bytes URLs in canonicalize_url()

### PR DESCRIPTION
**Pull request by redapple**
Fixes https://github.com/scrapy/scrapy/issues/2304

### CODE REVIEW

File: w3lib/url.py [hunk 1]
```diff
@@ -422,7 +422,9 @@ def parse_url(url, encoding=None):
     parsed url)
     """
     if isinstance(url, ParseResult):
-        return url
+        # we need to make sure we have unicode parts
+        # otherwise, things like IDNA encoding may fail on bytes
+        return ParseResult(*(to_unicode(part, encoding) for part in url))
```
Comment (kmike) [inline w3lib/url.py]: I wonder if we can remove the performance penalty from Python 3 code (if there is a penalty). 

Also, there is an issue with encodings - this assumes utf8 everywhere which may not be the case. Why is Scrapy urlparse-ing bytes urls? Should it use safe_ParseResult instead?

Comment (kmike) [inline w3lib/url.py]: It seems the root issue is that IDNA encoding is defined only for unicode, not for bytes. So we should convert netloc to unicode before encoding it in _safe_ParseResult.

Comment (redapple) [inline w3lib/url.py]: @kmike , regarding your first comment, scrapy is using [`to_native_string()` in the default lxml-based link extractor](https://github.com/scrapy/scrapy/blob/129421c7e31b89b9b0f9c5f7d8ae59e47df36091/scrapy/linkextractors/lxmlhtml.py#L59), which is not correct, and for which there's an open PR https://github.com/scrapy/scrapy/pull/1949

**Comment by Gallaecio** [comment]
Maybe we should close this? #2304 seems to have been fixed already.

**Comment by Gallaecio** [comment]
No longer applies, Scrapy dropped Python 2 support.

**Pull request closed without merging**